multisig-ism/src/main.sw

```diff
@@ -0,0 +1,125 @@
+contract;
+
+dep interface;
+dep metadata;
+
+use std::{vm::evm::{evm_address::EvmAddress, ecr::ec_recover_evm_address}, logging::log};
+
+use hyperlane_message::EncodedMessage;
+
+use merkle::StorageMerkleTree;
+
+use interface::MultisigIsm;
+
+use metadata::{MultisigMetadata, commitment};
+
+
+storage {
+    // TODO: consider u32 => struct
+    threshold: StorageMap<u32, u8> = StorageMap {},
+    validators: StorageMap<u32, Vec<EvmAddress>> = StorageMap {},
+    commitment: StorageMap<u32, b256> = StorageMap {},
+}
+
+#[storage(read, write)]
+fn update_commitment(domain: u32) {
+    let validators = storage.validators.get(domain);
+    let threshold = storage.threshold.get(domain);
+    storage.commitment.insert(domain, commitment(threshold, validators));
+}
+
+#[storage(read)]
+fn is_enrolled(domain: u32, validator: EvmAddress) -> bool {
+    let validators = storage.validators.get(domain);
+    let mut i = 0;
+    let len = validators.len();
+    while i < len {
+        if validators.get(i).unwrap() == validator {
+            return true;
+        }
+        i += 1;
+    }
+    return false;
+}
+
+#[storage(read)]
+fn validators(domain: u32) -> Vec<EvmAddress> {
+    storage.validators.get(domain)
+}
+
+impl MultisigIsm for Contract {
+    #[storage(read)]
+    fn threshold(domain: u32) -> u8 {
+        storage.threshold.get(domain)
+    }
+
+    #[storage(read)]
+    fn validators(domain: u32) -> Vec<EvmAddress> {
+        validators(domain)
+    }
+
+    #[storage(read)]
+    fn is_enrolled(domain: u32, validator: EvmAddress) -> bool {
+        is_enrolled(domain, validator)
+    }
+
+    #[storage(read)]
+    fn verify(metadata: MultisigMetadata, message: EncodedMessage) -> bool {
+        let calculated_root = StorageMerkleTree::branch_root(message.id(), metadata.proof, metadata.index);
+        require(metadata.root == calculated_root, "!merkle");
+
+        let origin = message.origin();
+        require(metadata.commitment() == storage.commitment.get(origin), "!commitment");
+
+        require(metadata.threshold <= metadata.signatures.len(), "!threshold");
```

Comment:
hmm interesting

Reply:
Actually I guess this doesn't matter because we're not receiving the metadata as bytes, we're receiving a struct. But in any case because we'll eventually move to opaque Bytes being passed in as metadata, this still kinda applies